sgkit/distance/core.py

```diff
@@ -0,0 +1,163 @@
+import typing
+import warnings
+
+import dask.array as da
+import numpy as np
+
+from sgkit.typing import ArrayLike
+
+
+def pairwise(
+    x: ArrayLike,
+    map_fn: np.ufunc,
+    reduce_fn: np.ufunc,
+    n_map_param: int,
+    chunks: typing.Optional[typing.Tuple[int, int]] = None,
+) -> ArrayLike:
+    """A generic pairwise function for any separable distance metric.
+    This calculates the pairwise distance between a set of vectors in the
+    given two-dimensional array, using the map reduce ufuncs. To illustrate
+    the algorithm consider the following (4, 5) two dimensional array:
```

Comment:
The example array is (4, 4), so need to either change the example or the stated shape. Maybe change the example to be non-square, so it's clear whether you are treating rows or columns as vectors to be compared?